### PR TITLE
Support integrity algorithm 128-EIA3

### PR DESCRIPTION
**Pull request by jinjiu**
REFERENCES:
1. https://www.gsma.com/security/wp-content/uploads/2019/05/EEA3_EIA3_specification_v1_8.pdf
2. https://www.gsma.com/security/wp-content/uploads/2019/05/eea3eia3testdatav11.pdf

### CODE REVIEW

File: crypto/eia3/eia3.c [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+    ZUC_init(zk);
+
+    return 1;
+}
+
+int EIA3_Update(EIA3_CTX *ctx, const unsigned char *inp, size_t len)
+{
+    ZUC_KEY *zk = &ctx->zk;
+    int i, remain, length = len * 8, num = ctx->num;
```
[the remark this answers is not in the document]
Comment (jinjiu) [reply crypto/eia3/eia3.c]: fixed.

File: crypto/eia3/eia3_pmeth.c [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+    EVP_MD_CTX_set_flags(mctx, EVP_MD_CTX_FLAG_NO_INIT);
+    EVP_MD_CTX_set_update_fn(mctx, int_update);
+
+    if (!ASN1_OCTET_STRING_set(&pctx->key, key->data, key->length))
+        return 0;
+
+    return EIA3_Init(&pctx->ctx, key->data, NULL);
+}
+static int eia3_signctx(EVP_PKEY_CTX *ctx, unsigned char *sig, size_t *siglen,
```
Comment (InfoHunter) [inline crypto/eia3/eia3_pmeth.c:117]: 这行之前加空行

Comment (jinjiu) [reply crypto/eia3/eia3_pmeth.c:118]: fixed

File: crypto/eia3/eia3_ameth.c [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    return 1;
+}
+
+const EVP_PKEY_ASN1_METHOD eia3_asn1_meth = {
+    EVP_PKEY_EIA3,
+    EVP_PKEY_EIA3,
+    0,
+
+    "EIA3",
+    "OpenSSL EIA3 method",
```
Comment (InfoHunter) [inline crypto/eia3/eia3_ameth.c:95]: 去掉OpenSSL字符

**Review by InfoHunter — APPROVED** [review]

File: crypto/eia3/eia3_ameth.c (hunk 3 above)
Comment (jinjiu) [reply crypto/eia3/eia3_ameth.c]: fixed

File: include/crypto/zuc.h [hunk 4]
```diff
@@ -36,8 +37,14 @@ typedef struct ZUC_KEY_st {
     int inited;
 } ZUC_KEY;
 
+typedef struct eia3_context EIA3_CTX;
+
 void ZUC_init(ZUC_KEY *zk);
 int ZUC_generate_keystream(ZUC_KEY *zk);
 void ZUC_destroy_keystream(ZUC_KEY *zk);
 
+int EIA3_Init(EIA3_CTX *ctx, const unsigned char *key, const unsigned char *iv);
```
Comment (dongbeiouba) [inline include/crypto/zuc.h:46]: What is length of key and iv?

Maybe you can refer to Poly1305，something like:

`void Poly1305_Init(POLY1305 *ctx, const unsigned char key[32]);`

File: crypto/eia3/eia3_local.h [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * Licensed under the BabaSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://github.com/BabaSSL/BabaSSL/blob/master/LICENSE
+ */
+
+#ifndef OPENSSL_NO_ZUC
+
+#include <stdlib.h>
```
Comment (dongbeiouba) [inline crypto/eia3/eia3_local.h:12]: Please add spaces for alignment

Comment (jinjiu) [reply crypto/eia3/eia3_local.h]: fixed

**Review by InfoHunter — APPROVED** [review]
It seems this is ready to be merged.

**Comment by InfoHunter** [comment]
Ping @dongbeiouba for a 2nd review

File: include/crypto/zuc.h (hunk 4 above)
Comment (jinjiu) [reply include/crypto/zuc.h]: fixed

File: crypto/eia3/eia3.c [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+#ifndef OPENSSL_NO_ZUC
+
+# include "crypto/zuc.h"
+# include "eia3_local.h"
+
+static ossl_inline uint32_t GET_WORD(uint8_t *data, uint32_t i)
+{
+    uint32_t word = 0, ti, j = i / 8;
+
+    ti  = i % 8;
```
[the remark this answers is not in the document]
Comment (dongbeiouba) [reply crypto/eia3/eia3.c:23]: ping @jinjiu

**Review by dongbeiouba — APPROVED** [review]

**Pull request merged**